src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -0,0 +1,472 @@
+import org.broadinstitute.hellbender.cmdline.argumentcollections.OptionalIntervalArgumentCollection;
+import org.broadinstitute.hellbender.cmdline.programgroups.DataFlowProgramGroup;
+import org.broadinstitute.hellbender.engine.dataflow.DataflowCommandLineProgram;
+import org.broadinstitute.hellbender.engine.dataflow.ReadsSource;
+import org.broadinstitute.hellbender.tools.dataflow.transforms.MarkDuplicatesReadsKey;
+import org.broadinstitute.hellbender.utils.GenomeLocSortedSet;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.dataflow.DataflowUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import static org.broadinstitute.hellbender.tools.dataflow.GenomicsReadUtils.getAlignmentStart;
+import static org.broadinstitute.hellbender.tools.dataflow.GenomicsReadUtils.getReferenceName;
+import static org.broadinstitute.hellbender.tools.dataflow.GenomicsReadUtils.isPrimaryAlignment;
+
+@CommandLineProgramProperties(
+        usage="Marks duplicates on dataflow",
+        usageShort="Mark Duplicates",
+        programGroup = DataFlowProgramGroup.class)
+public final class MarkDuplicatesDataflow extends DataflowCommandLineProgram {
+
+    //Bases below this quality will not be included in picking the best read from a set of duplicates.
+    private static final int MIN_BASE_QUAL = 15;
+
+    @Argument(doc="a prefix for the dataflow output files", shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
+    protected String outputFile;
+
+    @Argument(doc = "uri for the input bam, either a local file path or a gs:// bucket path",
+            shortName = StandardArgumentDefinitions.INPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.INPUT_LONG_NAME,
+            optional = false)
+    protected String bam;
+
+    @ArgumentCollection
+    protected IntervalArgumentCollection intervalArgumentCollection = new OptionalIntervalArgumentCollection();
+
+    @Override
+    protected void setupPipeline(final Pipeline pipeline) {
+        final ReadsSource readsSource = new ReadsSource(bam, pipeline);
+        final SAMFileHeader header = readsSource.getHeader();
+        final SAMSequenceDictionary sequenceDictionary = header.getSequenceDictionary();
+        final List<SimpleInterval> intervals = intervalArgumentCollection.intervalsSpecified() ? intervalArgumentCollection.getIntervals(sequenceDictionary):
+                getAllIntervalsForReference(sequenceDictionary);
+
+        final PCollectionView<SAMFileHeader> headerPcolView = pipeline.apply(Create.of(header)).apply(View.<SAMFileHeader>asSingleton());
+
+        final PCollection<Read> preads = readsSource.getReadPCollection(intervals);
```

Comment:
Here and elsewhere, we'll need to switch to the Read interface. If David's refactor doesn't get in first, this can get merged as-is.